# Conditionals Loops and Functions

---

## Learning Objectives

*  Understand and apply various types of Conditional Statements in practical examples.
*  Understand and utilize the different types of Loops in Python.
*  Integrate Conditional Statements with Loops.
*  Develop and use Functions with parameters and arguments.

---

## 1.0 Conditional Statements

### `if`,`else` and `elif`

**Conditional statements** like "**if, else, and elif**" help your code make decisions. They let you choose which part of your code to run based on whether certain conditions are met (**True**) or not met (**False**).
It is very essential for writing codes that can adapt to different situations.

### `if` statements

**if** statements are the most basic form of conditional statements. They run a block of code only when a specified condition is met.

###  Nested `if` statements

Nested **if** statements help make decisions that involve multiple conditions.

### `if`-`else` statements

An **if-else** statement runs one set of instructions if a condition is `True`, and another set of instructions if it's `False`.

## 1.1 Practical Example

<div align="center">
<h4>Figure 1: Anti Deforestation Campaign</h4>
<img src="https://c7.alamy.com/comp/2D69TY6/environmental-activists-campaign-against-deforestation-vector-flat-cartoon-illustration-of-protesting-eco-activists-with-posters-on-demonstration-sa-2D69TY6.jpg" width="50%">
</div>

#### I. Issue Deforestation Warning

Write a code **using the `if` statement** to issue a warning when deforestation exceeds a certain level.

In [ ]:
# Check if the deforestation rate exceeds 20% and print a warning if it does.
deforestation_rate = 0.25  # ie 25%
if deforestation_rate > 0.2:
    print("Critical deforestation level reached!")

Critical deforestation level reached!


#### II. Issue Urgent Action Notice Over Deforestation of Protected Areas

Write a code **using a nested `if` statement** that checks if the deforestation rate exceeds a certain level in an area classified as a `protected area`, and issues a warning if it does.

In [ ]:
# Check if the deforestation rate exceeds 15% in a protected area and print an urgent action message if it does.

# Step 1: Define both variables
deforestation_rate = 0.18
protected_area = True

# Step 2: Apply conditional statements
if deforestation_rate > 0.15:
    if protected_area:  # "if protected_area:" is equivalent to "if protected_area == True"
        print("Urgent action needed in protected area due to high deforestation rate!")

Urgent action needed in protected area due to high deforestation rate!


#### III. Issue Deforestation Alert Notification

Write a code **using the `if-else` statements** to send an alert message if deforestation activity is detected.

In [ ]:
# Print a deforestation alert message if the alert is active, otherwise indicate no alerts.
deforestation_alert = True

if deforestation_alert:
    print("Alert: Deforestation activity detected!")
else:
    print("No deforestation alerts.")

Alert: Deforestation activity detected!


#### IV. Classify Deforestation Levels

Write a code **using the `elif` statement** to classify the different levels of deforestation detected based on 3 degrees of deforestation: `"Critical Level"`, `"High Level"`, and `"Normal Level"`.

In [ ]:
# Classify and print the deforestation rate as 'Critical level', 'High level', or 'Normal level'
# based on its percentage.
deforestation_rate = 0.1

if deforestation_rate > 0.2:
    print("Critical level")
elif deforestation_rate > 0.1:
    print("High level")
else:
    print("Normal level")

Normal level


### Chained Conditional Statements

**Chained conditional statements** use `elif` to check multiple conditions. They are often useful when you need to choose exactly **one outcome** from several possibilities.

For example, you could use them to decide on different environmental actions to take based on the severity of deforestation detected.

In [ ]:
# Determine and print appropriate environmental actions based on the current deforestation rate level.
deforestation_rate = 0.21

if deforestation_rate > 0.3:
    print("Implement emergency measures")
elif deforestation_rate > 0.2:
    print("Increase surveillance and conservation efforts")
elif deforestation_rate > 0.1:
    print("Monitor and plan reforestation")
else:
    print("Deforestation rate is within safe limits")

Increase surveillance and conservation efforts


#### How It Works

When the code runs, it checks the stated conditions in order, stopping as soon as it finds a condition that holds `True` (ie the condition is met). Even if multiple conditions are met, only the first condition met is executed.

## 1.2 Conditional Statements Keywords

In Python, keywords within conditional statements are used for more flexibility in our code.

`if not`: Checks if a condition is **not true**.

`pass`: The pass keyword does nothing. It is only used as a placeholder when a **statement is needed but no action is required**.

For example, suppose we want to create an automated alert system for forest conservation that tells us the forest if healthy if no deforestation is detected, and alerts us if that is not the case.

In [ ]:
# Prints "Forest is healthy!" if no deforestation is detected, otherwise alerts that deforestation has been detected.
deforestation_detected = True

if not deforestation_detected:
    print("Forest is healthy! 😊")
else:
    print("Deforestation detected! 😧")

Deforestation detected! 😧


Now suppose we want to create a system that monitors deforestation levels and prompts us to take action only when necessary. However, we are yet to decide on what the line of action should be if deforestation exceeds a predefined level.

An important thing to note is that an empty `if` block of code is invalid in Python. You must always have at least one intended output instruction. In this case, we can use the `pass` statement as our output instruction.

In [ ]:
# This code will result in an error because the 'if' block is empty.
deforestation_rate = 0.05

if deforestation_rate > 0.1:
else:
    print("No immediate action required")

IndentationError: expected an indented block after 'if' statement on line 4 (3025528076.py, line 5)

In [ ]:
# Placeholder for future alert implementation if deforestation rate exceeds 10%, otherwise prints
# no immediate action is needed.
deforestation_rate = 0.1

if deforestation_rate > 0.1:
    pass  # Future implementation of alert system
else:
    print("No immediate action required")

No immediate action required


## 1.3 Ternary Operator

**Ternary operators** are a shortcut for `if-else` statements. They let us choose between two options based on a condition in one line of code.

For example, you could use a ternary operator to quickly decide if deforestation is at a `"High"` or `"Normal"` based on a value.

In [ ]:
# Sets and prints the deforestation status based on the deforestation rate.
deforestation_rate = 0.5
status = "High" if deforestation_rate > 0.1 else "Normal"

print(f"Deforestation status: {status}")

Deforestation status: High


---

## 2.0 Loops

**Loops** are a fundamental concept in Python programming. They are used to repeat tasks.

For example, we could use a loop to check if each animal in a list of animals in a conservation park is endangered. In this case, we have a list of animals and we want to print them out one by one. We could use "list indexing" to achieve this. However, if we have a large list of say 100 animals, we would have to write a print statement for each animal This is a very repetitive task and an inefficient way of programming prone to errors.

In [7]:
# List of animals (both endangered and not)
animals = ['Great White Shark', 'Blue Whale', 'African Elephant',  'Bald Eagle', 'Orangutan', 'Tiger', 'Panda', 'Koala']

print(animals[0])
print(animals[1])
print(animals[2])
print(animals[3])

# etc.

Great White Shark
Blue Whale
African Elephant
Bald Eagle


Loops make it easy to repeat actions more efficiently.


* `for` loops are perfect for iterating through a list, tuple, or string.


* `while` loops are used to keep iterating for as long as a condition is **True**.


**Note**

To **iterate** in programming means to repeatedly execute a set of instructions. We can iterate through a list or tuple in Python.


### `for` loop

In [2]:
list_of_numbers =[0,1,2,3,4,5, 6]
for i in list_of_numbers:  # Loop body
    print(i)

0
1
2
3
4
5
6


### `while` loop

In [6]:
i = 0

while i < 5:
    print(i)
    i += 1

0
1
2
3
4


## 2.1 Syntax and Structure of Loops

### `for` loop

The syntax of a `for` loop in Python is straightforward. It includes the **for keyword**, an **item variable**, the **in operator**, and a **collection**, where collection refers to an object we can iterate through, like a list, dictionary, or tuple.

The instructions we want to run in each loop are in the loop body.

In [8]:
for animal in animals: # animals list defined in an earlier line of code
    print(animal)

Great White Shark
Blue Whale
African Elephant
Bald Eagle
Orangutan
Tiger
Panda
Koala


### **Note**

You can put as many instructions as you want inside a loop. When the loop is finished, the code goes on to the next line after the loop.

In [9]:
animals_we_have_seen = [] #This is an empty list of animals we have encountered

for animal in animals:
    print(animal)
    animals_we_have_seen.append(animal) #add the animal to the "seen" list.

print(f" We saw these animals:\n {animals_we_have_seen}")

Great White Shark
Blue Whale
African Elephant
Bald Eagle
Orangutan
Tiger
Panda
Koala
 We saw these animals:
 ['Great White Shark', 'Blue Whale', 'African Elephant', 'Bald Eagle', 'Orangutan', 'Tiger', 'Panda', 'Koala']


Here is a breakdown of the above code in simple terms:

**Creates an empty list**: The code starts by making a list called `animals_we_have_seen`. This list is initially empty.

**Looks at each animal**: It then iterates through the list of animals for each animal.

**Prints the animal**: It prints the name of the animal for each iteration.

**Adds to the seen animals list**: It adds the animal it has iterated to to the `animals_we_have_seen` list.

**Shows the seen animals list**: It prints all the animals it has seen/iterated through.

Basically, the code goes through a list of animals, prints each one and creates a list to keep track of the animals it has seen.

### **Note**

We can loop through any iterable data structure like `lists`, `tuples`, `sets` and `dictionaries`.

In [10]:
# Define the tuple and set

animals_tuple = ('Great White Shark', 'Blue Whale', 'African Elephant',  'Bald Eagle', 'Orangutan', 'Tiger', 'Tiger', 'Panda', 'Koala')
animals_set= set(animals_tuple)

In [ ]:
print("Tuple:")
for animal in animals_tuple:
    print(animal)

Tuple:
Great White Shark
Blue Whale
African Elephant
Bald Eagle
Orangutan
Tiger
Tiger
Panda
Koala


In [ ]:
print("Set:")
for animal in animals_set:
    print(animal)

Set:
Bald Eagle
Tiger
Orangutan
African Elephant
Panda
Blue Whale
Koala
Great White Shark


### `while` loop

A `while` loop in Python is used to repeatedly execute its body for as long as a defined condition is **True**. Its syntax is given as:

while condition:
    # Loop body

`while` loops are ideal for scenarios where you do not know how many times you need to iterate through a collection.

For example, adding 10% to a value until it is >5. It is difficult to ascertain how many times we have to repeat this simple calculation. However, with a `while` loop, we can keep iterating until we get the value for which our condition is met.

In [13]:
population = 1 #We start at 0

while population <= 5:
    print(population) # We print out the value of population
    population *= 1.1 # And add 10% to it (ie 1+0.1 = 1.1)

1
1.1
1.2100000000000002
1.3310000000000004
1.4641000000000006
1.6105100000000008
1.771561000000001
1.9487171000000014
2.1435888100000016
2.357947691000002
2.5937424601000023
2.853116706110003
3.1384283767210035
3.4522712143931042
3.797498335832415
4.177248169415656
4.594972986357222


The `while` loop keeps printing different values of population with 10% added to each iteration until the population is 4.59 and it stops because the next iteration's value does not meet the set condition (ie population <= 5).

## 2.2 Loop Controls

**Loop control** statements let you change how loops work. They provide flexibility and control over how loops behave in different scenarios. The primary loop control statements in Python are; `break`, `continue`, and `pass`.

### `break` statement

It is often used when searching for a particular item in a list. Once that item is found, the `break` statement is used to stop the loop.

In [14]:
# Searching for a specific number in a list
animals = ['Great White Shark', 'Blue Whale', 'African Elephant',  'Bald Eagle', 'Orangutan', 'Tiger', 'Panda', 'Koala']

for animal in animals:
    print(animal)
    if animal == 'African Elephant':
        print("African Elephant found")
        break #Once African Elephant is found, the loop stops, so 'Bald Eagle', 'Orangutan', 'Tiger', 'Panda', 'Koala' are not printed.

print(f"The last animal checked = {animal}")

Great White Shark
Blue Whale
African Elephant
African Elephant found
The last animal checked = African Elephant


Notice that the loop stopped once it got to the African Elephant animal and then proceeded to the next line of code outside the loop to print the last animal checked.

### `continue` statement

The `continue` statement, on the other hand, skips the rest of the loop body for the current/ongoing iteration and moves to the next iteration. It comes in very handy when you want to skip/ filter out certain elements while looping through the collection.

In [15]:
for animal in animals:

    if animal == 'African Elephant':
        continue
    print(animal)

#Once the African Elephant is found, the code in the rest of the loop is not executed, so the African Elephant found is not printed.

Great White Shark
Blue Whale
Bald Eagle
Orangutan
Tiger
Panda
Koala


---

## 3,0 Python Functions

### What is a Function?

A **`function`** is a block of **organized reusable code** that is used in python to perform an action.

There are 2 basic types of functions that we can work with in Python.
* Built-in Functions.
* User-defined Functions.

### Built-in Functions

Python offers a variety of built-in functions that have been constructed on our behalf; which we can quickly and easily utilize.

Follow the **[link](https://docs.python.org/3/library/functions.html)** to access a list of all the Python built-in functions we have at our disposal.


Built-in functions come with a set of allowances and constraints that must be obeyed in order for the function to perform properly. These rules are listed in what we call the **`documentation`** of the function. This is like the guideline for what the function is and how to use it. To access the documentation of a function, we can use Python's **`help`** utility.

### User-defined Functions

<div align="center" style=" font-size: 80%; text-align: center; margin: 0 auto">
<img src="https://raw.githubusercontent.com/Explore-AI/Pictures/master/function_components.png"  style="width:70%";/>
<br>
<br>
    <em>Figure 2: Components of a function in Python</em>
</div>

Here is how to create a function in Python:

* **Start with def**:Use the keyword **`def`** to indicate you want to define a function.

* **Name your function**: Choose a descriptive name that reflects the function's purpose. Avoid using reserved keywords or names that are already defined in Python.

* **List inputs (optional)**: If the function needs information, list its inputs (also known as parameters) inside parentheses. The inputs act as placeholder for the actual values that will be passed to the function whenever you call it.

* **Write the code**: Write the code that you want the function to run. Make sure to indent the code underneath the function. This code block defines the functions use/behaviour.

* **Return a value**: Use the return statement to specify the value that the function should return.

## 3.1 Practical Example

Building on our previous examples, let us attempt to write a function named **`calc_forest_area`**. This function should take in a length and width as input, and **calculate the area** of a given forest by multiplying both input values.

### Area of Forest Function

#### Step I

Define the function and its arguments.

Write the code to pass the desired instructions to the function.

Return a value as the output.

In [16]:
#create the calc_forest_area function
def calc_forest_area(length, width): #define the function and the arguments
    area = length * width            #function instructions
    return area                      #return statement

#### Step II

Call the function

In [17]:
calc_forest_area(750, 9000)

6750000

You can store your output in a variable to use it later.

In [19]:
zambesi_forest = calc_forest_area(300, 5000)


1500000


### Key Points To Remember

* Functions can take any number of inputs or none at all.

* Functions can return any type of value, or they can return `None`.

* Functions can be called multiple times from different parts of your code.

* By using functions, you can break down your code into smaller more manageable pieces, making it easier to understand, debug and maintain.

## 3.2 Scope

**Scope** determines where a variable can be used in your Python code.

**Local variables** are created inside functions, and are only accessible within those functions. This means that if you try to access a local variable outside its function, you will get an error. This is because local variables have a limited scope and their existence is confined to the function where they are defined.

**Global variables** are created outside functions and can be accessed from anywhere in your code. This means that you can use a global variable both inside and outside functions. However, it's generally considered good practice to use global variables sparingly, as they can make your code harder to understand and maintain.

<div align="center">
<img src=https://miro.medium.com/v2/resize:fit:720/format:webp/1*1dIPs62xthLfoWCwwq0Vug.jpeg width="60%">
<h4>Figure 3: Global vs Local Variables In Python</h4>
</div>

Here is a simple example to illustrate the difference between local and global variables:

In [24]:
# Global variable
global_variable = 10

def my_function():
    # Local variable
    local_variable = 20

    # Accessing local variable from within the function
    print("Local variable inside the function:", local_variable)

    # Accessing global variable from within the function
    #print("Global variable inside the function:", global_variable)

    # Trying to access a local variable outside the function (will raise an error)
#print("Local variable outside the function:", local_variable)

# Calling the function
my_function()

# Accessing global variable outside the function
print("Global variable outside the function:", global_variable)

Local variable inside the function: 20
Global variable outside the function: 10


In the example above, we defined a `global_variable` that can be accessed both inside and outside the `my_function` function. On the other hand, `local_variable` is a variable that can only be accessed within the `my_function` function. If you try to access the `local_variable` outside the function, you will get an error.

### Key Points To Remember

* Local variables are created within functions and have a limited scope.

* Global variables are created outside functions and can be accessed from anywhere in your code.
